File: test/language/service_types/java/service_types/simple_service/SimpleServiceTest.java
package service_types.simple_service;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;
import java.math.BigInteger;

import org.junit.jupiter.api.Test;

import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.Writer;
import zserio.runtime.service.ServiceData;
import zserio.runtime.service.ServiceException;

public class SimpleServiceTest
{
    @Test
    public void serviceFullName()
    {
        assertEquals("service_types.simple_service.SimpleService",
                SimpleService.SimpleServiceService.serviceFullName());
    }

    @Test
    public void methodNames()
    {
        assertEquals("powerOfTwo", SimpleService.powerOfTwo_METHOD_NAME);
        assertEquals("powerOfTwo", SimpleService.SimpleServiceService.methodNames()[0]);
    }

    @Test
    public void powerOfTwo()
    {
        final Request request = new Request(13);
        assertEquals(BigInteger.valueOf(169), client.powerOfTwoMethod(request).getValue());
        request.setValue(-13);
        assertEquals(BigInteger.valueOf(169), client.powerOfTwoMethod(request).getValue());
        request.setValue(2);
        assertEquals(BigInteger.valueOf(4), client.powerOfTwoMethod(request).getValue());
        request.setValue(-2);
        assertEquals(BigInteger.valueOf(4), client.powerOfTwoMethod(request).getValue());
    }

    @Test
    public void powerOfTwoRaw() throws IOException
    {
        final Request request = new Request(13);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        request.write(writer);
        final byte[] requestData = writer.toByteArray();
        final byte[] responseData = client.powerOfTwoRawMethod(requestData);
        final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(responseData);
        final Response response = new Response(reader);
        assertEquals(BigInteger.valueOf(169), response.getValue());
    }

    @Test
    public void invalidServiceMethod()
    {
        assertThrows(ServiceException.class, () -> service.callMethod("nonexistentMethod", null, null));
    }

    @Test
    public void callWithContext()
    {
        final FakeContext fakeContext = new FakeContext();
        assertFalse(fakeContext.seenByService);
        final Request request = new Request(10);
        final Response response = client.powerOfTwoMethod(request, fakeContext);
        assertEquals(BigInteger.valueOf(100), response.getValue());
        assertTrue(fakeContext.seenByService);
    }

    private static class Service extends SimpleService.SimpleServiceService
    {
        @Override
        public Response powerOfTwoImpl(Request request, Object context)
        {
            if (context != null)
                ((FakeContext)context).seenByService = true;

            final Response response = new Response(BigInteger.valueOf(request.getValue()).pow(2));
            return response;
        }

        @Override
        public byte[] powerOfTwoRawImpl(byte[] requestData, Object context)
        {
            try
            {
                final ByteArrayBitStreamReader reader = new ByteArrayBitStreamReader(requestData);
                final Request request = new Request(reader);
                final Response response = powerOfTwoImpl(request, context);
                final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
                response.write(writer);
                return writer.toByteArray();
            }
            catch (IOException e)
            {
                return new byte[] {};
            }
        }
    }

    private static class FakeContext
    {
        public boolean seenByService = false;
    }

    private static class LocalServiceClient implements zserio.runtime.service.ServiceClientInterface
    {
        public LocalServiceClient(Service service)
        {
            this.service = service;
        }

        @Override
        public byte[] callMethod(String methodName, ServiceData<? extends Writer> request, Object context)
        {
            return service.callMethod(methodName, request.getByteArray(), context).getByteArray();
        }

        private final Service service;
    }

    private static final Service service = new Service();
    private static final LocalServiceClient localServiceClient = new LocalServiceClient(service);
    private static final SimpleService.SimpleServiceClient client =
            new SimpleService.SimpleServiceClient(localServiceClient);
}

File: test/extensions/language/service_types/cpp/SimpleServiceTest.cpp
#include <fstream>
#include <memory>

#include "gtest/gtest.h"
#include "service_types/simple_service/SimpleService.h"
#include "test_utils/LocalServiceClient.h"
#include "zserio/RebindAlloc.h"
#include "zserio/SerializeUtil.h"
#include "zserio/ServiceException.h"

using namespace zserio::literals;

namespace service_types
{
namespace simple_service
{

using allocator_type = SimpleService::Client::allocator_type;
using LocalServiceClient = test_utils::LocalServiceClient<allocator_type>;
template <typename T>
using vector_type = zserio::vector<T, allocator_type>;
using BitBuffer = zserio::BasicBitBuffer<allocator_type>;

namespace
{
struct FakeContext
{
    bool seenByService = false;
};
} // namespace

class SimpleServiceImpl : public SimpleService::Service
{
public:
    Response powerOfTwoImpl(const Request& request, void* context) override
    {
        if (context != nullptr)
        {
            static_cast<FakeContext*>(context)->seenByService = true;
        }

        int32_t value = request.getValue();
        if (value < 0)
        {
            value = -value;
        }
        return Response(static_cast<uint64_t>(value) * static_cast<uint64_t>(value), get_allocator_ref());
    }

    vector_type<uint8_t> powerOfTwoRawImpl(zserio::Span<const uint8_t> requestData, void* context) override
    {
        zserio::BitStreamReader reader(requestData.data(), requestData.size());
        Request request(reader);
        Response response = powerOfTwoImpl(request, context);

        vector_type<uint8_t> responseData;
        responseData.resize((response.bitSizeOf() + 7) / 8);
        zserio::BitStreamWriter writer(responseData.data(), responseData.size());
        response.write(writer);

        return responseData;
    }
};

class SimpleServiceTest : public ::testing::Test
{
public:
    SimpleServiceTest() :
            localServiceClient(service),
            client(localServiceClient),
            movedService(createService()),
            movedLocalServiceClient(movedService),
            movedClient(createClient())
    {}

private:
    // check that move constructors work correctly
    SimpleServiceImpl createService()
    {
        // cannot be const because we need to move-construct it
        SimpleServiceImpl createdService;
        return createdService;
    }

    SimpleService::Client createClient()
    {
        // cannot be const because we need to move-construct it
        SimpleService::Client createdClient(movedLocalServiceClient);
        return createdClient;
    }

protected:
    SimpleServiceImpl service;
    LocalServiceClient localServiceClient;
    SimpleService::Client client;
    SimpleServiceImpl movedService;
    LocalServiceClient movedLocalServiceClient;
    SimpleService::Client movedClient;
};

TEST_F(SimpleServiceTest, serviceFullName)
{
    ASSERT_EQ("service_types.simple_service.SimpleService"_sv, SimpleService::Service::serviceFullName());
}

TEST_F(SimpleServiceTest, methodNames)
{
    ASSERT_EQ("powerOfTwo"_sv, SimpleService::Service::methodNames()[0]);
}

TEST_F(SimpleServiceTest, powerOfTwo)
{
    Request request;

    request.setValue(13);
    Response response = client.powerOfTwoMethod(request);
    ASSERT_EQ(169, response.getValue());
    request.setValue(-13);
    response = movedClient.powerOfTwoMethod(request);
    ASSERT_EQ(169, response.getValue());
    request.setValue(2);
    response = client.powerOfTwoMethod(request);
    ASSERT_EQ(4, response.getValue());
    request.setValue(-2);
    response = movedClient.powerOfTwoMethod(request);
    ASSERT_EQ(4, response.getValue());
}

TEST_F(SimpleServiceTest, powerOfTwoRaw)
{
    Request request;
    request.setValue(13);
    auto requestData = zserio::serialize(request);
    auto responseData = client.powerOfTwoRawMethod(requestData.getBytes());
    auto response = zserio::deserialize<Response>(BitBuffer(responseData));
    ASSERT_EQ(169, response.getValue());
}

TEST_F(SimpleServiceTest, invalidServiceMethod)
{
    ASSERT_THROW(service.callMethod("nonexistentMethod"_sv, {}, nullptr), zserio::ServiceException);
}

TEST_F(SimpleServiceTest, callWithContext)
{
    FakeContext fakeContext;
    ASSERT_FALSE(fakeContext.seenByService);
    Request request{10};
    Response response = client.powerOfTwoMethod(request, &fakeContext);
    ASSERT_EQ(100, response.getValue());
    ASSERT_TRUE(fakeContext.seenByService);

    fakeContext.seenByService = false;
    response = movedClient.powerOfTwoMethod(request, &fakeContext);
    ASSERT_EQ(100, response.getValue());
    ASSERT_TRUE(fakeContext.seenByService);
}

} // namespace simple_service
} // namespace service_types
